framework/sandstone.cpp

```diff
@@ -2611,70 +2610,93 @@ static int exec_mode_run(int argc, char **argv)
 {
     using Socket = Topology::Package;
 
     int orig_verbosity;
     Topology topo = Topology::topology();
-    vector<Socket>::iterator it;
-    vector<int> disabled_sockets;
     vector<int> result; // faulty sockets
     int ret = EXIT_SUCCESS;
     bool ever_failed = false;
 
-    it = topo.packages.begin();
-    if (topo.packages.empty())
-        return result;                  // shouldn't happen!
-
-    if (topo.packages.size() == 1) {
-        result.push_back(it->id);
-        return result;
-    }
-
     // backup the original verbosity
     orig_verbosity = sApp->verbosity;
     sApp->verbosity = -1;
 
-    for (; it != topo.packages.end(); disabled_sockets.push_back(it->id), ++it) {
-        vector<Socket>::iterator eit = it;
+    // first "non-empty" package found in the topology
+    vector<Socket>::iterator first_it = topo.packages.end();
+    // first "non-empty" package being "removed" from the list
+    vector<Socket>::iterator removed_it = topo.packages.end();
+    // last "non-empty" package being added
+    vector<Socket>::iterator last_it = topo.packages.end();
+
+    vector<Socket>::iterator it = topo.packages.begin();
+    while (it != topo.packages.end()) {
+        vector<Socket>::iterator eit;
         LogicalProcessorSet run_cpus = {};
-        int k = 0;
 
-        for (; eit != topo.packages.end(); ++eit)
-            run_cpus.add_package(*eit);
+        // merge all CPUs from all packages in the list (w/ respect to the removed packages)
+        for (eit = it; eit != topo.packages.end(); ++eit) {
+            if (!(*eit).cores.empty()) {
+                if (first_it == topo.packages.end()) {
+                    first_it = eit;
+                }
+                last_it = eit;
+                run_cpus.add_package(*eit);
+            }
+        }
 
-        sApp->enabled_cpus = run_cpus;
-        do {
-            ret = run_tests_on_cpu_set(triage_tests, run_cpus);
-        } while (!ret && ++k < sApp->retest_count);
+        if (!run_cpus.empty()) {
+            sApp->enabled_cpus = run_cpus;
+            // should we run any tests if retest_count == 0??
+            // Keep the loop running only for SUCCESSes
+            int k = 0;
+            do {
+                ret = run_tests_on_cpu_set(triage_tests, run_cpus);
+            } while ((ret == EXIT_SUCCESS) && ++k < sApp->retest_count);
 
-        if (ret) ever_failed = true; /* we've seen a failure */
+            if (ret > EXIT_SUCCESS) ever_failed = true; /* we've seen a **FAILURE** */
 
-        if (!ret && ever_failed) {
-            // the last socket removed is the main suspect
-            result.push_back(disabled_sockets.at(disabled_sockets.size() - 1));
-            break;
+            if ((ret == EXIT_SUCCESS) && ever_failed) {
+                // the last socket removed is the main suspect
+                result.push_back((*removed_it).id);
```

Comment:
Thinking... can we get `removed_it == end` here? By construction, we can't get here in the first loop iteration, because we can either set `ever_failed` or get here. So if this is the second loop, we must have run the code below.

Reply:
No, "removed" entry is set properly (excluded conditions). At least second set must be handled to get here (first with FAIL, second with SUCCESS).
I considered to replace 'ever_failed' with 'removed != end()'.. but I wasn't sure of this change (and therefore left it). Will replace.